### PR TITLE
MAINT: Update `openblas_support.py` to support Apple Silicon

### PR DESCRIPTION
**Pull request by judahrand**
#### What does this implement/fix?
This commit brings `scipy`'s implementation of this script into line with `numpy`'s. This includes an update of the OpenBlas version as well as support for `macosx-arm64`.
    
The source implementation can be found here: https://github.com/numpy/numpy/blob/7ef28b2f1c383828bebcbb87b8e487e27c6e3ff9/tools/openblas_support.py

I believe this is needed before the checks on https://github.com/MacPython/scipy-wheels/pull/123 will pass.

### CODE REVIEW

**Review by tylerjereddy — COMMENTED** [review]
I'm assuming you've read and dealt with the issues described here: https://github.com/scipy/scipy/issues/14188

**Comment by judahrand** [comment]
> I'm assuming you've read and dealt with the issues described here: #14188

I apologise, I did not! I was coming at this from a different perspective (from `MacPython/scipy-wheels`) and also not being familiar with the SciPy release system etc. Should have searched issues first.

**Comment by tylerjereddy** [comment]
Ok, I don't know what to do about that set of issues, but maybe it will get more attention now that M1 support is gaining momentum.

**Comment by rgommers** [comment]
Not quite sure that that slowdown reported on gh-14188 is real. This PR:

image

Another recent PR (`#14284)`:

image

Yet another recent PR (`#14308`):

image

**Comment by rgommers** [comment]
There's also much faster runs - timings jump around a lot.

**Comment by judahrand** [comment]
@rgommers Either way it looks like `0.3.15` might not be far off so perhaps worth waiting for that and including here?

**Comment by judahrand** [comment]
@rgommers The 0.3.15 libs seem to be available now anyway so have bumped to them which hopefully avoids any issues.

**Comment by isuruf** [comment]
To fix linux-32bit test, try merging https://github.com/scipy/scipy/pull/14333 into this branch.

**Comment by isuruf** [comment]
Now the issue is that `gfortran-5` is used in the test. Need to change that to `gfortran-8`.

**Comment by rgommers** [comment]
> Now the issue is that `gfortran-5` is used in the test. Need to change that to `gfortran-8`.

Why is this the problem? The only failure I see is `numpy.distutils.system_info.NotFoundError: No BLAS/LAPACK libraries found.`, and that shouldn't be related to `gfortran` version.

We need one test with an old gfortran, otherwise we'll run into it at wheel build time. xref https://github.com/MacPython/scipy-wheels/pull/111 which got stuck, we'd like to bump the minimum `gfortran` version, but `8.x` is definitely too high (`5.5` or `6.x` seems like a good bump).

**Comment by isuruf** [comment]
Actual error is,
```
/usr/bin/ld: warning: libgfortran.so.5, needed by /usr/lib/gcc/i686-linux-gnu/5/../../../../lib/libopenblas.so, not found (try using -rpath or -rpath-link)
/usr/lib/gcc/i686-linux-gnu/5/../../../../lib/libopenblas.so: undefined reference to `_gfortran_concat_string@GFORTRAN_8'
/usr/lib/gcc/i686-linux-gnu/5/../../../../lib/libopenblas.so: undefined reference to `_gfortran_etime@GFORTRAN_8'
collect2: error: ld returned 1 exit status
```

That's because openblas-0.3.15 was built with gfortran 8 and you need `libgfortran`

**Comment by rgommers** [comment]
Ah yes, I managed to overlook that build log output, thanks.

> That's because openblas-0.3.15 was built with gfortran 8 and you need `libgfortran>=8`

Hmm. I looked through the PRs on https://github.com/MacPython/openblas-libs and https://github.com/MacPython/openblas-libs. There's a bunch of upgrades, but it's not entirely clear to me if this was a necessary change.

If we can make our wheels work with `gfortran>=8`, then we can take an ATLAS build for testing an older `gfortran`. I'm not sure we can get away with `>=8` given `manylinux` standards and `pip` versions out in the wild though.

**Comment by isuruf** [comment]
Linux 32bit and windows 32 bit both fail with the same error now.

**Comment by judahrand** [comment]
https://github.com/scipy/scipy/issues/14188#issuecomment-870910279

Would bumping the tolerances of that test be an acceptable solution as suggested?

**Comment by isuruf** [comment]
I'm not sure bumping tolerance will help as the differences are huge, but if you can get it to work with a bump in tolerance I think that's okay.

**Comment by isuruf** [comment]
Any thoughts on marking the test as xfail on 32 bit systems? I doubt anybody cares about interpolative decomposition on 32-bit systems.

**Comment by tylerjereddy** [comment]
> Any thoughts on marking the test as xfail on 32 bit systems? I doubt anybody cares about interpolative decomposition on 32-bit systems.

@ilayn what do you think about that idea?

**Comment by rgommers** [comment]
Re-running to see the actual failures. Unfortunately 32-bit Windows is fairly heavily used, so maybe we should raise an exception in the relevant functionality for 32-bit systems if it's niche rather than return wrong answers.

**Comment by ilayn** [comment]
This is a bit indeed tricky. 

Two things we know for sure, Microsoft stopped shipping 32-bit windows of any kind together with 2020 Win10 2004 update. Also Nvidia stopped supporting 32-bit systems a few years back. So there are no new users coming in but there are some still out there. However, I don't think we can stop and smell the flowers at this point for the rest of the userbase. On the other hand not everybody is privileged to jump on a new system as years go by. So this is one of those cases that make you hate being an open source maintainer 😃 

I am not sure if this is relevant but can we cross-compile for 32-bit systems with gfortran8? Or would that even help?

**Comment by isuruf** [comment]
> I am not sure if this is relevant but can we cross-compile for 32-bit systems with gfortran8? Or would that even help? 

I don't know how that would help at all

> Because I don't see why this diverges this much. I'll check if some ordering has changed inside the array

Interpolative decomposition is returning a rank of 5 instead of 19. Rank 5 is way too low to reconstruct the hilbert matrix and that's why we see a large error.

**Comment by matthew-brett** [comment]
For Windows 32-bit - I can't check immediately - but is it still true that the Python.org Windows default binary installer is 32-bit?    That's the question at hand - how many 32-bit Python installations there are out there, nearly all of which will be on 64-bit Windows.  My guess is that there are still a lot of them.  @zooba - do you have any update on this?  It's a question that comes up fairly often.

**Comment by isuruf** [comment]
> That's the question at hand - how many 32-bit Python installations there are out there, nearly all of which will be on 64-bit Windows. 

I think the real question is how many users are using `scipy.linalg.interpolative_decompisition` on 32-bit windows. I'd guess None.

**Comment by zooba** [comment]
We switched the default install for 3.9 to be 64-bit, so the number of 32-bit installs will decrease. However, I believe Anaconda is still offering both at equal footing (though I guess you aren't too concerned about the build there, but it may influence source changes).

Until we get Windows ARM64 support (which I don't expect you to be able to do right now), the 32-bit build is the one that those users will need. It's likely that there are already more of them than those on a genuine 32-bit operating system.

And it's always going to be the case that the 32-bit build of CPython will use less memory than the 64-bit one. For large multi-node workloads that aren't going to exceed 2GB of allocation individually, that can be a serious advantage. And in general, we all get very little insight into those kinds of workloads.

Ultimately, it's probably best to widely announce deprecation and see who cries, which is a really sucky way to do it, but I can't think of a better way. At the very least, that's a good threshold for "how painful is it to keep 32-bit running?" 😉

**Comment by isuruf** [comment]
There are two changes in this PR. Update openblas to 0.3.15 and update gfortran from 5 to 8. It's the gfortran update from 5 to 8 that's the issue here. cc @inducer

**Comment by judahrand** [comment]
> maybe we should raise an exception in the relevant functionality for 32-bit systems if it's niche rather than return wrong answers.

When you say 'the relevant functionality' what do we mean? The whole `linalg.interpolative` module? Or a subset of the functions because the way the test is written currently it is pretty hard to workout what the broken subset might be. Is it all of `interp_decomp`? Is it `reconstruct_matrix_from_id`? Or is a lot of everything broken?

Do you think the best first step is to break apart `test_id` into more useful tests?

**Comment by rgommers** [comment]
> When you say 'the relevant functionality' what do we mean? The whole `linalg.interpolative` module? Or a subset of the functions because the way the test is written currently it is pretty hard to workout what the broken subset might be. Is it all of `interp_decomp`? Is it `reconstruct_matrix_from_id`? Or is a lot of everything broken?

The subset that is broken. And yes, it's hard to know what that subset is.

> Do you think the best first step is to break apart `test_id` into more useful tests?

That'd probably be easier, such a PR can be merged straight away and then this PR can be rebased on top of it. Could I interest you in making that PR? (I know you uncovered a bit of a mess here, probably more than you initially signed up for so please feel free to say no)

**Comment by judahrand** [comment]
> > When you say 'the relevant functionality' what do we mean? The whole `linalg.interpolative` module? Or a subset of the functions because the way the test is written currently it is pretty hard to workout what the broken subset might be. Is it all of `interp_decomp`? Is it `reconstruct_matrix_from_id`? Or is a lot of everything broken?
> 
> The subset that is broken. And yes, it's hard to know what that subset is.
> 
> > Do you think the best first step is to break apart `test_id` into more useful tests?
> 
> That'd probably be easier, such a PR can be merged straight away and then this PR can be rebased on top of it. Could I interest you in making that PR? (I know you uncovered a bit of a mess here, probably more than you initially signed up for so please feel free to say no)

Having a go at it now. Will try to get a PR in this evening (almost 10pm here).

**Comment by judahrand** [comment]
Do you also think it is worth bumping OpenBLAS further to v0.3.17? I notice Numpy has in the latest release.

**Comment by inducer** [comment]
There's a potentially newer version of the ID software that's being distributed here: http://tygert.com/software.html It might be worth trying to see if that already contains a fix.

**Comment by rgommers** [comment]
> There's a potentially newer version of the ID software that's being distributed here: http://tygert.com/software.html It might be worth trying to see if that already contains a fix.

Thanks @inducer. How do you see it's newer? It's just a tarball, closest I can find that gives a hint of how old it is is the date in the doc pdf. Which says March 2014, and you updated the version in SciPy in July 2014 in commit a9af5e3acab.

**Comment by mattip** [comment]
> Do you also think it is worth bumping OpenBLAS further to v0.3.17?

Yes.

**Comment by mattip** [comment]
Over at #14583, it seems updating win32 mingw to 7.3.0 is sufficient to pass tests.

**Comment by isuruf** [comment]
@inducer, @rgommers, I checked the tarball and the only differences between that and scipy are whitespace differences.

@judahrand, can you use gfortran 7 instead of 8 for both window and linux?

**Comment by judahrand** [comment]
> @inducer, @rgommers, I checked the tarball and the only differences between that and scipy are whitespace differences.
> 
> @judahrand, can you use gfortran 7 instead of 8 for both windows and linux?

Will give it a go when I get a chance

**Comment by rgommers** [comment]
Rebased on master after the test refactor and 32-bit skips for `complex128` were merged. Let's see if it's happier now.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -201,16 +201,16 @@ stages:
             python3.7 get-pip.py && \
             pip3 --version && \
             pip3 install setuptools wheel numpy==1.16.6 cython==0.29.18 pybind11 pytest pytest-timeout pytest-xdist pytest-env pytest-cov Pillow mpmath matplotlib==3.1.3 pythran --user && \
-            apt-get -y install gcc-5 g++-5 gfortran-5 wget && \
+            apt-get -y install gcc-5 g++-5 gfortran-7 wget && \
```
Comment (isuruf) [inline azure-pipelines.yml:204]:
```suggestion
            apt-get -y install gcc-5 g++-5 gfortran-8 wget && \
```

File: azure-pipelines.yml [hunk 2]
```diff
@@ -201,16 +201,16 @@ stages:
...
             cd .. && \
             mkdir openblas && cd openblas && \
             target=\$(python3.7 ../scipy/tools/openblas_support.py) && \
             cp -r \$target/lib/* /usr/lib && \
             cp \$target/include/* /usr/include && \
             cd ../scipy && \
-            CC=gcc-5 CXX=g++-5 F77=gfortran-5 F90=gfortran-5 python3.7 setup.py install && \
+            CC=gcc-5 CXX=g++-5 F77=gfortran-7 F90=gfortran-7 python3.7 setup.py install && \
```
Comment (isuruf) [inline azure-pipelines.yml:211]:
```suggestion
            CC=gcc-5 CXX=g++-5 F77=gfortran-8 F90=gfortran-8 python3.7 setup.py install && \
```

File: azure-pipelines.yml [hunk 3]
```diff
@@ -201,16 +201,16 @@ stages:
...
             python3.7 tools/openblas_support.py --check_version $(openblas_version) && \
-            CC=gcc-5 CXX=g++-5 F77=gfortran-5 F90=gfortran-5 python3.7 runtests.py --mode=full -- -n auto -s --junitxml=junit/test-results.xml --cov-config=.coveragerc --cov-report=xml --cov-report=html"
+            CC=gcc-5 CXX=g++-5 F77=gfortran-7 F90=gfortran-7 python3.7 runtests.py --mode=full -- -n auto -s --junitxml=junit/test-results.xml --cov-config=.coveragerc --cov-report=xml --cov-report=html"
```
Comment (isuruf) [inline azure-pipelines.yml:213]:
```suggestion
            CC=gcc-5 CXX=g++-5 F77=gfortran-8 F90=gfortran-8 python3.7 runtests.py --mode=full -- -n auto -s --junitxml=junit/test-results.xml --cov-config=.coveragerc --cov-report=xml --cov-report=html"
```

**Comment by judahrand** [comment]
> @judahrand, can you use gfortran 7 instead of 8 for both windows and linux?

Okay, shall revert these changes back to fortran-8.

File: azure-pipelines.yml [hunk 4]
```diff
@@ -18,10 +18,10 @@ pr:
     include:
     - '*'  # must quote since "*" is a YAML reserved character; we want a string
 
-# the version of OpenBLAS used is currently 0.3.8.dev
+# the version of OpenBLAS used is currently 0.3.15
 # and should be updated to match scipy-wheels as appropriate
 variables:
-    openblas_version: 0.3.9
+    openblas_version: 0.3.15
```
Comment (isuruf) [inline azure-pipelines.yml:24]:
```suggestion
    openblas_version: 0.3.17
```

**Comment by tylerjereddy** [comment]
The pre release azure failure should go away--I just merged in Ralf's fix for that a minute ago.

**Review by rgommers — APPROVED** [review]
All right, all green except for one unrelated CI issue that was fixed earlier today. It's time to merge this. Thanks a lot @judahrand and @isuruf!

**Comment by tylerjereddy** [comment]
Labelled for backport, I think

**Pull request merged**